lib/commands/context.js

```diff
@@ -0,0 +1,461 @@
+         !this.isRealDevice() &&
+         this.caps.safari;
+};
+
+extensions.navToInitialWebview = async function () {
+  let timeout = 0;
+  if (this.isRealDevice()) {
+    timeout = 3000;
+    logger.debug(`Waiting for ${timeout} ms before navigating to view.`);
+  }
+  await B.delay(timeout);
+  if (this.useNewSafari()) {
+    await this.typeAndNavToUrl();
+  } else if (parseInt(this.iosSdkVersion, 10) >= 7 && !this.isRealDevice() && this.caps.safari) {
+    await this.navToViewThroughFavorites();
+  } else {
+    await this.navToViewWithTitle(/.*/);
+  }
+};
+
+extensions.typeAndNavToUrl = async function () {
+  let initialUrl = this.caps.safariInitialUrl || `http://127.0.0.1:${this.caps.port}/welcome`;
+  let oldImpWait = this.implicitWaitMs;
+  this.implicitWaitMs = 7000;
+  let el = await this.findElement('name', 'URL');
+  this.implicitWaitMs = oldImpWait;
+  await this.nativeTap(el.ELEMENT);
+  let els = await this.findElements('name', 'Address');
+  let addressEl = els[els.length - 1].ELEMENT;
+  await this.setValueImmediate(addressEl, initialUrl);
+  el = await this.findElement('name', 'go');
+  await this.nativeTap(res.ELEMENT);
+  await this.navToViewWithTitle(/.*/i);
+  // Waits for page to finish loading.
+  await this.remote.pageUnload();
+};
+
+extensions.navToViewThroughFavorites = async function () {
+  logger.debug('We are on iOS7+ simulator: clicking apple button to get into a webview');
+  let oldImpWait = this.implicitWaitMs;
+  this.implicitWaitMs = 7000; // wait 7s for apple button to exist
+  try {
+    let el = await this.findElement('xpath', '//UIAScrollView[1]/UIAButton[1]');
+  } catch (err) {
+    let msg = 'Could not find button to click to get into webview. ' +
+              'Proceeding on the assumption we have a working one.';
+    logger.error(msg);
+    this.implicitWaitMs = oldImpWait;
+    return await this.navToViewWithTitle(/.*/i);
+  }
+  this.implicitWaitMs = oldImpWait;
+  try {
+    let res = await this.nativeTap(el.ELEMENT);
+    if (res.status !== 0) {
+      throw new Error('Failed!');
+    }
+  } catch (err) {
+    let msg = 'Could not click button to get into webview. ' +
+              'Proceeding on the assumption we have a working one.';
+    logger.error(msg);
+  }
+  await this.navToViewWithTitle(/apple/i);
+};
+
+extensions.navToViewWithTitle = async function (titleRegex) {
+  logger.debug('Navigating to most recently opened webview');
+  let start = Date.now();
+  let spinTime = 500;
+  let spinHandles = async () => {
+    let res;
+    try {
+      res = await this.getLatestWebviewContextForTitle(titleRegex);
+    } catch (err) {
+      throw new Error(`Could not navigate to webview! Err: ${err.message}`);
+    }
+    if (res) {
+      let latestWindow = res;
+      logger.debug(`Picking webview '${latestWindow}'`);
+      await this.setContext(latestWindow);
+      await this.remote.cancelPageLoad();
+      return;
+    } else {
+      if ((Date.now() - start) < 90000) {
```

Comment:
I think it would be more elegant if there were one try/catch in here and one `return await spinHandles()`